xof/xof.go

```diff
@@ -0,0 +1,86 @@
+// Package xof provides an interface for eXtendable-Output Functions.
+package xof
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/cloudflare/circl/internal/sha3"
+	"golang.org/x/crypto/blake2b"
+	"golang.org/x/crypto/blake2s"
+)
+
+// XOF defines the interface to hash functions that support arbitrary-length output.
+type XOF interface {
+	// Write absorbs more data into the hash's state. It panics if called
+	// after Read.
+	io.Writer
+
+	// Read reads more output from the hash. It returns io.EOF if the limit
+	// has been reached.
+	io.Reader
+
+	// Clone returns a copy of the XOF in its current state.
+	Clone() XOF
+
+	// Reset resets the XOF to its initial state.
+	Reset()
+}
+
+type ID uint
+
+func (x ID) Available() bool { return x < maxID && xofRegistry[x] != nil }
+func (x ID) XofIDFunc() ID   { return x }
+func (x ID) New() XOF {
+	if x < maxID {
+		f := xofRegistry[x]
+		if f != nil {
+			return f()
+		}
+	}
+	panic(fmt.Errorf("crypto: requested XOF function #%v is unavailable", x))
+}
+
+func RegisterXOF(x ID, f func() XOF) {
+	if x >= maxID {
+		panic("crypto: RegisterXOF of unknown XOF function")
+	}
+	xofRegistry[x] = f
+}
+
+var xofRegistry = make([]func() XOF, maxID)
+
+func init() {
```

Comment:
This makes this package which just defines an interface import these other packages, which makes the whole registration business a bit less useful. Not that the registration was very useful in the first place imho.

Reply:
agree, registry not needed at all, changed now.